LevelDB::DB.{new,create,load} should coerce the pathname argument into a string

DataWraith opened the pull request:
This patch is intended to match the behavior of `File.open` in order to be less surprising. `File.open` coerces the filename argument into a string before attempting to use it. This makes it possible to use the Pathname convenience-wrapper from the standard library.

Unfortunately, I'm not quite sure how File.open does it exactly, so this patch changes the type of exception LevelDB throws, while File.open will still throw a TypeError.

Before:

``` ruby
class Foo; end

LevelDB::DB.new(Foo)
# => TypeError: wrong argument type Class (expected String)

LevelDB::DB.new(Pathname.new('/tmp/foo'))
# => TypeError: wrong argument type Pathname (expected String)
```

After:

``` ruby
class Foo; end

LevelDB::DB.new(Foo)
# => NoMethodError: undefined method `to_str' on Foo (Class)

LevelDB::DB.new(Pathname.new('/tmp/foo'))
# => #<LevelDB::DB:0x6324 @pathname="/tmp/foo"> 
```

Commit:
Call #to_str on the argument given to DB#{new, create, load}

This matches the behavior of `File.open`. In particular, it allows us
to pass Pathname objects (from the pathname standard library) to the
DB functions.

DataWraith commented:
Hm. After digging into this some more, it seems that Pathname#to_str has been [removed](http://redmine.ruby-lang.org/issues/show/1970) since 1.9 :-/

Apparently the correct solution for 1.9 would be to use the new File.path(pathname) instead of #to_str. File.path automatically handles pathnames via the new Pathname#to_path. This is not backwards-compatible with 1.8 though, so handling all cases would probably require a helper like this:

``` ruby
def get_path(pathname)
  return pathname if pathname.is_a?(String)
  return pathname.to_path if pathname.respond_to?(:to_path)
  return pathname.to_str if pathname.respond_to?(:to_str)
  raise TypeError, "Could not coerce #{pathname.inspect} to String"
end
```

At this point, it breaks the beautiful simplicity of the current implementation, and it's probably not worth caring about at all. Ah well.

ghost commented:
Why not use `#to_s`?

``` ruby
ruby-1.9.2-p290 :001 > class Foo
ruby-1.9.2-p290 :002?>   def to_s
ruby-1.9.2-p290 :003?>     "wakka wakka"
ruby-1.9.2-p290 :004?>   end
ruby-1.9.2-p290 :005?> end
 => nil 
ruby-1.9.2-p290 :006 > String(Foo.new)
 => "wakka wakka" 
ruby-1.9.2-p290 :007 >
```

`#to_str` was an alias of `#to_s` anyway.

DataWraith commented:
#to_str and #to_s differ semantically, see http://briancarper.net/blog/98/, especially the comment by Grant Hutchins.

Apart from the fact, that the intent was to match the native functions (such as File.open and Kernel.open), which use #to_str in 1.8 and File.path in 1.9, the problem I see with #to_s is that it is defined for just about everything, because it is defined in Object. For example: 

``` ruby
nil.to_s   # => ""
Array.to_s # => "Array"
{:foo => :bar }.to_s # => "foobar"
```

This makes it less obvious when you pass something wrong into a method by accident.

wmorgan commented:
I'm ok with an explicit check. Something like:

``` ruby
    def to_path x
      File.respond_to?(:path) ? File.path(x) : x.to_str
    end
```

doesn't seem overly clumsy.

DataWraith commented:
Neat, I hadn't thought of doing it the other way around, and testing File. I'll add a new commit adding a #path_string helper method; #to_path might be confusing, because by convention it returns the object it is called _on_ as String.

Commit:
Use File.path to coerce pathname arguments into Strings in 1.9

The previous commit introduced the use of #to_str on the given pathname
to coerce it into a String, matching the behavior of File.open and other
file handling functions under 1.8.

Ruby 1.9 introduces the File.path method for this purpose and deprecates
the use of #to_str on classes such as Pathname. This commit introduces
a private helper, path_string, that uses File.path when available and
falls back to #to_str otherwise.

Tested with MRI 1.8.7 and 1.9.2.

wmorgan commented:
Not to keep harping on something that already works, but I wonder whether:
  Pathname.new(x).to_s

would be an even nicer solution. Seems to work on both 1.8 and 1.9, when x is both a string and a Pathname itself.

DataWraith commented:
That's a possibility I hadn't thought of; you'd just need to `require 'pathname'`. 

It can even be shortened to `Pathname(x).to_s` and does seem to match the `File.open` behavior with regard to type coercion.

On the other hand, it feels a little hackish. Strictly speaking, `File.path(x)` is the "correct" method to use in 1.9, and `x.to_str` in 1.8, but  Pathname(x).to_s is much clearer in intent, so I definitively wouldn't mind if you changed it.

wmorgan commented:
Some web page somewhere claimed that this was the normal approach in Ruby stdlibs. I'm not sure if that's true. But I'm not inclined to change it unless someone complains. Just wanted to document it.
